File: backend/GraficaModerna.Application/Services/CartService.cs
using GraficaModerna.Application.DTOs;
using GraficaModerna.Application.Interfaces;
using GraficaModerna.Domain.Entities;
using GraficaModerna.Domain.Interfaces;
using Microsoft.Extensions.Logging;
using System.Data;

namespace GraficaModerna.Application.Services;

public class CartService(IUnitOfWork uow, ILogger<CartService> logger) : ICartService
{
    private const int MaxConcurrencyRetries = 3;
    private const int MaxQuantityPerItem = 5000;
    private readonly IUnitOfWork _uow = uow;
    private readonly ILogger<CartService> _logger = logger;

    public async Task<CartDto> GetCartAsync(string userId)
    {
        var cart = await GetOrCreateCart(userId);

        var itemsDto = cart.Items
            .Where(i => i.Product != null && i.Product.IsActive)
            .Select(i => new CartItemDto(
                i.Id,
                i.ProductId,
                i.Product!.Name,
                (i.Product.ImageUrls != null && i.Product.ImageUrls.Count > 0) ? i.Product.ImageUrls[0] : "",
                i.Product.Price,
                i.Quantity,
                i.Product.Price * i.Quantity,
                i.Product.Weight,
                i.Product.Width,
                i.Product.Height,
                i.Product.Length
            )).ToList();

        var orphanedItems = cart.Items.Where(i => i.Product == null || !i.Product.IsActive).ToList();
        if (orphanedItems.Count != 0)
        {
            foreach (var item in orphanedItems) await _uow.Carts.RemoveItemAsync(item);
            await _uow.CommitAsync();
        }

        return new CartDto(cart.Id, itemsDto, itemsDto.Sum(i => i.TotalPrice));
    }

    public async Task AddItemAsync(string userId, AddToCartDto dto)
    {
        ArgumentNullException.ThrowIfNull(dto);
        if (string.IsNullOrWhiteSpace(userId))
            throw new ArgumentException("userId inválido.", nameof(userId));

        if (dto.Quantity <= 0)
            throw new ArgumentOutOfRangeException(nameof(dto), "Quantidade deve ser maior que zero.");

        if (dto.Quantity > MaxQuantityPerItem)
            throw new ArgumentOutOfRangeException(nameof(dto),
                $"Quantidade excede o limite permitido de {MaxQuantityPerItem}.");

        for (var attempt = 1; attempt <= MaxConcurrencyRetries; attempt++)
        {
            using var transaction = await _uow.BeginTransactionAsync(IsolationLevel.ReadCommitted);
            try
            {
                var product = await _uow.Products.GetByIdWithLockAsync(dto.ProductId)
                    ?? throw new InvalidOperationException("Produto indisponível ou removido.");

                if (product.StockQuantity < dto.Quantity)
                    throw new InvalidOperationException("Estoque insuficiente para a quantidade solicitada.");

                var cart = await _uow.Carts.GetByUserIdWithLockAsync(userId);

                if (cart == null)
                {
                    cart = new Cart
                    {
                        UserId = userId,
                        LastUpdated = DateTime.UtcNow
                    };
                    await _uow.Carts.AddAsync(cart);
                    await _uow.CommitAsync();
                }

                var existing = cart.Items.FirstOrDefault(i => i.ProductId == dto.ProductId);

                if (existing != null)
                {
                    long newTotal = (long)existing.Quantity + dto.Quantity;

                    if (newTotal > MaxQuantityPerItem)
                        throw new InvalidOperationException(
                            $"O total de itens excederia o limite máximo de {MaxQuantityPerItem}.");

                    if (newTotal > int.MaxValue)
                        throw new InvalidOperationException("Quantidade inválida (Excesso de itens).");

                    if (product.StockQuantity < newTotal)
                        throw new InvalidOperationException(
                            "Não é possível adicionar mais itens: estoque insuficiente.");

                    existing.Quantity = (int)newTotal;
                }
                else
                {
                    cart.Items.Add(new CartItem
                    {
                        CartId = cart.Id,
                        ProductId = dto.ProductId,
                        Quantity = dto.Quantity
                    });
                }

                cart.LastUpdated = DateTime.UtcNow;
                await _uow.CommitAsync();
                await transaction.CommitAsync();

                _logger.LogInformation(
                    "Item adicionado ao carrinho. User: {UserId}, Product: {ProductId}, Qty: {Quantity}",
                    userId, dto.ProductId, dto.Quantity);

                return;
            }
            catch (Exception ex)
            {
                await transaction.RollbackAsync();

                var isConcurrencyException = ex.GetType().Name.Contains("Concurrency") || ex.GetType().Name.Contains("DbUpdateConcurrency");

                if (isConcurrencyException)
                {
                    _logger.LogWarning(ex,
                        "Tentativa {Attempt} de {MaxAttempts} falhou (concorrência). User: {UserId}",
                        attempt, MaxConcurrencyRetries, userId);

                    if (attempt == MaxConcurrencyRetries)
                        throw new InvalidOperationException(
                            "Não foi possível processar a operação devido a alta concorrência. Tente novamente.");

                    await Task.Delay(TimeSpan.FromMilliseconds(100 * attempt));
                }
                else
                {
                    throw;
                }
            }
        }
    }

    public async Task UpdateItemQuantityAsync(string userId, Guid cartItemId, int quantity)
    {
        if (string.IsNullOrWhiteSpace(userId))
            throw new ArgumentException("userId inválido.", nameof(userId));

        if (quantity < 0)
            throw new ArgumentException("A quantidade não pode ser negativa.");

        if (quantity > MaxQuantityPerItem)
            throw new ArgumentException($"Quantidade excede o limite permitido de {MaxQuantityPerItem}.");

        if (quantity == 0)
        {
            await RemoveItemAsync(userId, cartItemId);
            return;
        }

        for (var attempt = 1; attempt <= MaxConcurrencyRetries; attempt++)
        {
            using var transaction = await _uow.BeginTransactionAsync(IsolationLevel.ReadCommitted);
            try
            {
                var cart = await _uow.Carts.GetByUserIdWithLockAsync(userId)
                    ?? throw new InvalidOperationException("Carrinho não encontrado.");

                var item = cart.Items.FirstOrDefault(i => i.Id == cartItemId)
                    ?? throw new InvalidOperationException("Item não encontrado no carrinho.");

                var product = await _uow.Products.GetByIdWithLockAsync(item.ProductId)
                    ?? throw new InvalidOperationException("Produto indisponível.");

                if (product.StockQuantity < quantity)
                    throw new InvalidOperationException("Estoque insuficiente.");

                item.Quantity = quantity;
                cart.LastUpdated = DateTime.UtcNow;

                await _uow.CommitAsync();
                await transaction.CommitAsync();
                return;
            }
            catch (Exception ex)
            {
                await transaction.RollbackAsync();

                var isConcurrencyException = ex.GetType().Name.Contains("Concurrency") || ex.GetType().Name.Contains("DbUpdateConcurrency");

                if (isConcurrencyException)
                {
                    if (attempt == MaxConcurrencyRetries)
                        throw new InvalidOperationException(
                            "Não foi possível atualizar o item devido a alta concorrência.");

                    await Task.Delay(TimeSpan.FromMilliseconds(100 * attempt));
                }
                else
                {
                    throw;
                }
            }
        }
    }

    public async Task RemoveItemAsync(string userId, Guid itemId)
    {
        if (string.IsNullOrWhiteSpace(userId))
            throw new ArgumentException("userId inválido.", nameof(userId));

        var cart = await GetOrCreateCart(userId);
        var item = cart.Items.FirstOrDefault(i => i.Id == itemId);

        if (item != null)
        {
            await _uow.Carts.RemoveItemAsync(item);
            await _uow.CommitAsync();
        }
    }

    public async Task ClearCartAsync(string userId)
    {
        if (string.IsNullOrWhiteSpace(userId))
            throw new ArgumentException("userId inválido.", nameof(userId));

        var cart = await _uow.Carts.GetByUserIdAsync(userId);
        if (cart != null)
        {
            await _uow.Carts.ClearCartAsync(cart.Id);
            await _uow.CommitAsync();
        }
    }

    private async Task<Cart> GetOrCreateCart(string userId)
    {
        if (string.IsNullOrWhiteSpace(userId))
            throw new ArgumentException("userId inválido.", nameof(userId));

        var cart = await _uow.Carts.GetByUserIdAsync(userId);
        if (cart == null)
        {
            cart = new Cart
            {
                UserId = userId,
                LastUpdated = DateTime.UtcNow
            };
            await _uow.Carts.AddAsync(cart);
            await _uow.CommitAsync();
        }

        return cart;
    }
}

File: app/Services/CartService.php
<?php

namespace App\Services;

use App\Models\Cart;
use App\Models\CartItem;
use App\Models\Product;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use Exception;

class CartService
{
    const MAX_QUANTITY_PER_ITEM = 5000;

    public function getCart(string $userId): Cart
    {
        // GetOrCreate logic igual ao C#
        $cart = Cart::firstOrCreate(
            ['user_id' => $userId],
            ['last_updated' => now()]
        );

        // Carrega items e produtos para montar o DTO
        $cart->load('items.product');

        // Lógica de Órfãos (Produtos nulos ou inativos) - C# Linhas 29-34
        $orphanedItems = $cart->items->filter(function ($item) {
            return $item->product === null || !$item->product->is_active;
        });

        if ($orphanedItems->isNotEmpty()) {
            CartItem::destroy($orphanedItems->pluck('id'));
            $cart->refresh(); // Recarrega o carrinho limpo
            // Commit é automático no final da request do Laravel, diferente do C# explícito
        }

        return $cart;
    }

    public function addItem(string $userId, array $dto): void
    {
        if (empty($userId)) throw new Exception("userId inválido.");
        if ($dto['quantity'] <= 0) throw new Exception("Quantidade deve ser maior que zero.");
        if ($dto['quantity'] > self::MAX_QUANTITY_PER_ITEM) 
            throw new Exception("Quantidade excede o limite permitido de " . self::MAX_QUANTITY_PER_ITEM . ".");

        // Loop de retry do C# é abstraído pelo segundo parâmetro do transaction (3 tentativas)
        DB::transaction(function () use ($userId, $dto) {
            
            // GetByIdWithLockAsync (C#) -> lockForUpdate (Laravel)
            $product = Product::where('id', $dto['productId'])->lockForUpdate()->first();

            if (!$product || !$product->is_active) 
                throw new Exception("Produto indisponível ou removido.");

            if ($product->stock_quantity < $dto['quantity'])
                throw new Exception("Estoque insuficiente para a quantidade solicitada.");

            $cart = Cart::firstOrCreate(
                ['user_id' => $userId],
                ['last_updated' => now()]
            );

            // Bloqueia o item do carrinho se existir
            $existingItem = CartItem::where('cart_id', $cart->id)
                ->where('product_id', $dto['productId'])
                ->lockForUpdate() 
                ->first();

            if ($existingItem) {
                $newTotal = $existingItem->quantity + $dto['quantity'];

                if ($newTotal > self::MAX_QUANTITY_PER_ITEM)
                    throw new Exception("O total de itens excederia o limite máximo de " . self::MAX_QUANTITY_PER_ITEM . ".");

                // Em PHP int overflow vira float, mas verificamos lógica
                if ($product->stock_quantity < $newTotal)
                    throw new Exception("Não é possível adicionar mais itens: estoque insuficiente.");

                $existingItem->quantity = $newTotal;
                $existingItem->save();
            } else {
                CartItem::create([
                    'cart_id' => $cart->id,
                    'product_id' => $dto['productId'],
                    'quantity' => $dto['quantity']
                ]);
            }

            $cart->update(['last_updated' => now()]);

            Log::info("Item adicionado ao carrinho. User: {$userId}, Product: {$dto['productId']}, Qty: {$dto['quantity']}");

        }, 3); // 3 tentativas (MaxConcurrencyRetries)
    }

    public function updateItemQuantity(string $userId, string $itemId, int $quantity): void
    {
        if (empty($userId)) throw new Exception("userId inválido.");
        if ($quantity < 0) throw new Exception("A quantidade não pode ser negativa.");
        if ($quantity > self::MAX_QUANTITY_PER_ITEM)
            throw new Exception("Quantidade excede o limite permitido de " . self::MAX_QUANTITY_PER_ITEM . ".");

        if ($quantity == 0) {
            $this->removeItem($userId, $itemId);
            return;
        }

        DB::transaction(function () use ($userId, $itemId, $quantity) {
            $cart = Cart::where('user_id', $userId)->first();
            if (!$cart) throw new Exception("Carrinho não encontrado.");

            $item = CartItem::where('id', $itemId)
                ->where('cart_id', $cart->id)
                ->first();
                
            if (!$item) throw new Exception("Item não encontrado no carrinho.");

            $product = Product::where('id', $item->product_id)->lockForUpdate()->first();
            
            if (!$product || !$product->is_active) throw new Exception("Produto indisponível.");

            if ($product->stock_quantity < $quantity)
                throw new Exception("Estoque insuficiente.");

            $item->quantity = $quantity;
            $item->save();
            
            $cart->update(['last_updated' => now()]);
        }, 3);
    }

    public function removeItem(string $userId, string $itemId): void
    {
        if (empty($userId)) throw new Exception("userId inválido.");

        $cart = $this->getCart($userId); // Reutiliza lógica de criar se não existir
        
        $item = $cart->items->where('id', $itemId)->first();

        if ($item) {
            $item->delete();
            $cart->touch();
        }
    }

    public function clearCart(string $userId): void
    {
        if (empty($userId)) throw new Exception("userId inválido.");

        $cart = Cart::where('user_id', $userId)->first();
        if ($cart) {
            $cart->items()->delete();
            $cart->touch();
        }
    }
}